Terminal.Gui/Views/ListView.cs

```diff
@@ -843,35 +879,64 @@ public Task SetSourceAsync (IList source)
     }
 
     private void ListView_LayoutStarted (object sender, LayoutEventArgs e) { EnsureSelectedItemVisible (); }
+    /// <summary>
+    /// Call the event to raises the <see cref="CollectionChanged"/>.
+    /// </summary>
+    /// <param name="e"></param>
+    protected virtual void OnCollectionChanged (NotifyCollectionChangedEventArgs e) { CollectionChanged?.Invoke (this, e); }
 }
 
 /// <summary>
 ///     Provides a default implementation of <see cref="IListDataSource"/> that renders <see cref="ListView"/> items
 ///     using <see cref="object.ToString()"/>.
 /// </summary>
-public class ListWrapper : IListDataSource
+public class ListWrapper<T> : IListDataSource
 {
-    private readonly int _count;
-    private readonly BitArray _marks;
-    private readonly IList _source;
+    private int _count;
+    private BitArray _marks;
+    private readonly ObservableCollection<T> _source;
 
     /// <inheritdoc/>
-    public ListWrapper (IList source)
+    public ListWrapper (ObservableCollection<T> source)
     {
         if (source is { })
         {
             _count = source.Count;
             _marks = new BitArray (_count);
             _source = source;
+            _source.CollectionChanged += (s, e) =>
+                                         {
+                                             CheckAndResizeMarksIfRequired ();
+                                             CollectionChanged?.Invoke (s, e);
+                                         };
             Length = GetMaxLengthItem ();
         }
     }
 
+    /// <inheritdoc />
+    public event NotifyCollectionChangedEventHandler CollectionChanged;
+
     /// <inheritdoc/>
-    public int Count => _source is { } ? _source.Count : 0;
+    public int Count => _source?.Count ?? 0;
 
     /// <inheritdoc/>
-    public int Length { get; }
+    public int Length { get; private set; }
+
+    void CheckAndResizeMarksIfRequired ()
+    {
+        if (_source != null && _count != _source.Count)
+        {
+            _count = _source.Count;
+            BitArray newMarks = new BitArray (_count);
```

Comment:
BitArray is _extremely_ expensive for what it does and involves a ton of extra heap activity with multiple copies at runtime.

It's good for compact storage - not for performance.

You'd get better performance with a literal List of bools or, better yet, since this is ephemeral, a stackalloced span of bools.

Reply:
There are also a lot of concurrency problems with the way this works as-is.

Reply:
Typically, ListView type controls are implemented via having the collection be a collection of items implementing a known interface, rather than a raw reference to a user-supplied collection.

That's how it's done in WPF and WinForms, for example.

Otherwise, the concurrency issues become a mess and it's hard to maintain good performance.

Reply:
I would say we should handle it like this:

Go ahead and merge this, just to give it the same behavioral starting point as the same change in v1.

But immediately file a new issue to address these problems, for v2.

Reply:
But this PR is for V2. For V1 it will be a breaking change. It will be better you explain with real or pseudo code.

Reply:
> There are also a lot of concurrency problems with the way this works as-is.

I'm only help with a lack of change collection event.

Reply:
If it's for letting as it was before then it's better leave to the user handle the necessary actions when he is manipulating the collection?

Reply:
It was always using `BitArray` from the beginning and `List<bool>` for some reason was not used. `Span<bool>` cannot be declared, as it gives the error below. I don't think it's worth changing this because `_marks` isn't even part of `_source`. The fact that it is handled in the `ListWrapper` does not mean that the `_source` changes. They are independent of each other. The only value `_marks` needs is `_source.Count`.

`CS8345 Field or self-implemented property cannot be of type 'Span<bool>' unless it is an instance member of a ref struct.`

Reply:
That's a red herring. Don't worry about the Span stuff for now. That has specific rules to follow, by virtue of being a ref struct, and pretty thorough understanding of how `Span<T>` and `Memory<T>` work is necessary to do that stuff without losing your sanity. 😅 

The best or at least most straightforward option with the fewest caveats is to wrap the collection, or else there's going to be a ton of work to do for concurrency on top of it.

The last suggestion I made was basically just to make them be the same and then we can fix it from there. That's mostly because my attention is on other things at the moment and I don't have the bandwidth to provide specific consultation on this particular part, right now, or else I'd be more than happy to give more detail or write something up myself. :)